lib/cache_swift.c

```diff
@@ -0,0 +1,518 @@
+/******************************************************************************
+ * $Id$
+ *
+ * Project:  MapServer
+ * Purpose:  MapCache tile caching support file: swift cache backend.
+ * Author:   Michael Downey and the MapServer team.
+ *
+ ******************************************************************************
+ * Copyright (c) 1996-2013 Regents of the University of Minnesota.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies of this Software or works derived from this Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *****************************************************************************/
+
+#include "mapcache.h"
+#ifdef USE_SWIFT
+
+#include "keystone-client.h"
+#include "swift-client.h"
+
+#include <apr_strings.h>
+#include <apr_reslist.h>
+#include <apr_hash.h>
+
+#include <string.h>
+#include <errno.h>
+#include <stdio.h>
+
+typedef struct mapcache_cache_swift mapcache_cache_swift;
+
+/**\class mapcache_cache_swift
+ * \brief a mapcache_cache for openstack Swift object storages using keystone authentication
+ * \implements mapcache_cache
+ */
+struct mapcache_cache_swift {
+   mapcache_cache cache;
+   char *auth_url;
+   char *tenant;
+   char *username;
+   char *password;
+   char *key_template;
+   char *container_template;
+};
+
+struct swift_conn_params {
+  mapcache_cache_swift *cache;
+};
+
+
+struct swift_connection {
+   keystone_context_t *keystone_context;
+   swift_context_t *swift_context;
+};
+
+void mapcache_swift_authenticate(mapcache_context *ctx, mapcache_cache_swift *cache, struct swift_connection *conn) {
+    enum keystone_error keystone_err;
+    enum swift_error swift_err;
+    char *url;
+    char *token;
+
+    keystone_err = keystone_authenticate(conn->keystone_context, cache->auth_url, cache->tenant, cache->username, cache->password);
+    if (keystone_err != KSERR_SUCCESS) {
+        ctx->set_error(ctx, 500, "failed to keystone_authenticate()");
```

Comment:
error conditions should halt execution. add a ```return``` here and after all ctx->set_error calls

Reply:
I will check all possible paths and fix these issues.